src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py

```diff
@@ -377,6 +378,66 @@ def list_key(cmd,
         raise CLIError(str(exception))
 
 
+def restore_key(cmd,
+                datetime,
+                key=None,
+                name=None,
+                label=None,
+                no_deletes=False,
+                connection_string=None,
+                yes=False):
+    current_retry = 0
+    max_retry = 3
+    retry_interval = 1
+
+    connection_string = resolve_connection_string(cmd, name, connection_string)
+    azconfig_client = AzconfigClient(connection_string)
+
+    query_option_then = QueryKeyValueCollectionOptions(key_filter=key,
+                                                       label_filter=QueryKeyValueCollectionOptions.empty_label if label is not None and not label else label,
+                                                       query_datetime=datetime)
+    query_option_now = QueryKeyValueCollectionOptions(key_filter=key,
+                                                      label_filter=QueryKeyValueCollectionOptions.empty_label if label is not None and not label else label)
+
+    try:
+        restore_keyvalues = azconfig_client.get_keyvalues(query_option_then)
+        current_keyvalues = azconfig_client.get_keyvalues(query_option_now)
+        kvs_to_restore, kvs_to_modify, kvs_to_delete = __compare_kvs_for_restore(restore_keyvalues, current_keyvalues, no_deletes)
+
+        if not yes:
+            need_change = __print_restore_preview(kvs_to_restore, kvs_to_modify, kvs_to_delete)
+            if need_change is False:
+                return
+
+        for kv in chain(kvs_to_restore, kvs_to_modify):
+            success = None
+            while success is None and current_retry <= max_retry:
+                try:
+                    success = azconfig_client.set_keyvalue(kv, ModifyKeyValueOptions())
```

Comment:
This API won't use etag to check precondition. In azconfig_client, there are three methods related to add/update kvs. 
1. add_keyvalue      use if-none-match:* to make sure adding a new kv
2. update_keyvalue       use if-match:etag to guarantee the kv we will change are based on what we get. 
3. set_keyvalue      this is a force set operation and doesn't check any precondition. 

In this case, we may want to use update_keyvalue

Reply:
I believe we discussed that we would do a force set operation. Let's say at time t1 the value is v1 and etag e1. At time t2, the value is updated to v2, with etag e2. What we are trying to do is set_keyvalue(k1, v1). If we use update_keyvalue(k1, v1, e1), this will fail, since the latest etag is e2. We could use e2 here but in any case, we had agreed that this should be a forced set since one value being changed in the meantime should not matter, since we are actually going to a time in the past.

Reply:
Yes, we should use force set kv. In this case, why we still catch precondition failed exception and keep retrying based on it?

Reply:
I borrowed that part form set_key. (I didn't use that because that does an additional retrieve which we don't need). On the other hand if you look at the other mass operation, import, there's no retry logic. I'm OK with either way, or a third option is retrying for all failures (not just precondition). Is there other precondition failures, other than etag that set_key could run into? Or is that check also erroneous?

Reply:
As for set_kv in line 182, it called either add_keyvalue or update_keyvalue which are both conditional. So we can retry based on precondition failed. However, for set_kv, it doesn't check precondition. For restore, it is fine that we don't retry. BTW, the data plane SDK will retry if it is retriable. On the other hand, since we restore sequentially, we need to consider the behavior for partial success. Either we fail as long as one kv failed or we just try best and error out kvs not restored.

Reply:
Agreed; removed the retry logic entirely. For partial success, I really believe we should fail where it fails, rather than going on with best effort. We are applying the changes in the same order they have confirmed and the user might not want to go on after a given failure. Also, there might be some bigger problems (server unavailable, connection, etc) which you might not want to try and fail for each and every key,